openshift/machine-api-operator
# Nodelink controller

How to test it:
- deploy k8s cluster with `aws-actuator bootstrap`: https://github.com/openshift/cluster-api-provider-aws/tree/master/cmd/aws-actuator#bootstrapping-cluster-api-stack
- build the nodelink controller: `make nodelink-controller`
- run it: `./bin/nodelink-controller --kubeconfig=KUBECONFIG`

Compared to the original version of the node-link controller (see https://github.com/openshift/cluster-operator/blob/master/pkg/controller/nodelink/node_link_controller.go), all the machines are cached. The original implementation checked only `kube-cluster` namespace. Given the node linking is needed for any machine and our autoscaler PoC relies on machineset that can be run in any namespaces, we need to iterate over all machines no matter in which namespace they are.

@frobware @dgoodwin @enxebre FYI

Sooner or later we will need to run the nodelink logic as part of an actuator. The current IP matching strategy is generic. Matching based on provider ID requires interpretation of a provider config which should be known only to its corresponding actuator.

We could even just set the node ref directly instead of annotating a node object.
